lib/services/android-project-service.ts

```diff
@@ -118,19 +120,134 @@ class AndroidProjectService implements IPlatformProjectService {
-		return (() => {
+        return (() => {
 			var buildConfiguration = options.release ? "release" : "debug";
-			var args = this.getAntArgs(buildConfiguration, projectRoot);
-			this.spawn('ant', args).wait();
+            var args = this.getAntArgs(buildConfiguration, projectRoot);
+            var args2 = this.getAntArgs(buildConfiguration, projectRoot);
+            this.spawn('ant', args).wait();
+            this.generateMetadata(projectRoot);
+            this.updateMetadata(projectRoot);
+            // build the project again in order to include the newly generated metadata
+            this.spawn('ant', args2).wait();
 		}).future<void>()();
 	}
 
 	public isPlatformPrepared(projectRoot: string): IFuture<boolean> {
 		return this.$fs.exists(path.join(projectRoot, "assets", constants.APP_FOLDER_NAME));
-	}
+    }
+
+    private generateBuildFile(projDir: string, targetSdk: string): void {
+        this.$logger.info("Generate build.xml for %s", projDir);
+        var cmd = util.format("android update project -p %s --target %s --subprojects", projDir, targetSdk);
+        this.$childProcess.exec(cmd).wait();
+    }
+
+    private parseProjectProperties(projDir: string, destDir: string): void {
+
+        var projProp = path.join(projDir, "project.properties");
+
+        if (!this.$fs.exists(projProp).wait()) {
+            this.$logger.warn("File %s does not exist", projProp);
+            return;
+        }
+
+        var lines = fs.readFileSync(projProp, { encoding: "utf-8" }).split(os.EOL);
+
+        lines.forEach((elem, idx, arr) => {
+            var match = elem.match(/android\.library\.reference\.(\d+)=(.*)/);
+            if (match) {
+                var libRef: ILibRef = { idx: parseInt(match[1]), path: match[2].trim() };
+                libRef.adjustedPath = this.$fs.isRelativePath(libRef.path) ? path.join(projDir, libRef.path) : libRef.path;
+                this.parseProjectProperties(libRef.adjustedPath, destDir);
+            }
+        });
+
+        this.$logger.info("Copying %s", projDir);
+        shell.cp("-Rf", projDir, destDir);
+
+        var targetDir = path.join(destDir, path.basename(projDir));
+        // TODO: parametrize targetSdk
+        var targetSdk = "android-17";
+        this.generateBuildFile(targetDir, targetSdk);
+    }
+
+    private getProjectReferences(projDir: string): ILibRef[]{
+        var projProp = path.join(projDir, "project.properties");
+
+        var lines = fs.readFileSync(projProp, { encoding: "utf-8" }).split(os.EOL);
+
+        var refs: ILibRef[] = [];
+
+        lines.forEach((elem, idx, arr) => {
+            var match = elem.match(/android\.library\.reference\.(\d+)=(.*)/);
+            if (match) {
+                var libRef: ILibRef = { idx: parseInt(match[1]), path: match[2] };
+                libRef.adjustedPath = path.join(projDir, libRef.path);
+                refs.push(libRef);
+            }
+        });
+
+        return refs;
```

Comment:
@fatme, your code can call result of calling ```null.map``` because match can return null. Imho, this is one of the cases, where the procedural code is shorter than the functional one. I would only get rid of the libref.adjustedPath line and move it in the returned object itself:

```
var refs: ILibRef[] = [];
lines.forEach(elem => {
	var match = elem.match(/android\.library\.reference\.(\d+)=(.*)/);
	if (match) {
            var libRef: ILibRef = { idx: parseInt(match[1]), path: match[2], adjustedPath: path.join(projDir, libRef.path) };
            refs.push(libRef);
        }
    });
return refs;
```